Split SDK into multiple projects (#24)
operator-framework/enhancements: enhancements/split-sdk-into-multiple-projects.md

---
title: split-sdk-into-multiple-projects
authors:
  - "@jesusr"
reviewers:
  - "@estroz"
  - "@hasbro17"
  - "@joelanford"
  - "@fabianvf"
  - "@shawn-hurley"
approvers:
  - "@estroz"
  - "@hasbro17"
  - "@joelanford"
  - "@fabianvf"
  - "@shawn-hurley"
creation-date: 2020-05-18
last-updated: 2020-05-27
status: implementable
see-also:
replaces:
superseded-by:
---

# Split SDK into multiple projects

## Release Signoff Checklist

- [X] Enhancement is `implementable`
- [X] Design details are appropriately documented from clear requirements
- [X] Test plan is defined
- [X] Graduation criteria for dev preview, tech preview, GA

## Open Questions [optional]

1. How will separate repos affect the SDK website and docs in general?
   * What if each individual repo maintains their own docs? Then have a
     pipeline sort of thing that would regenerate the entire doc if there are

>      any changes

Reviewer: We could build a mechanism to add docs from Ansible/Helm repos into the main sdk site at build time. This would keep the top level site as it is now. A drawback would be that the documentation versioning would be for the entire SDK, but it seems like the Ansible operator and its docs should be versioned separately. @ecordell posed a similar plan for pulling in all operator-framework docs into a single repo at build time, which could be used to version the framework as a single entity.
We have a few options, but they each have their drawbacks.
1. We could have separate sites for the ansible/helm sdks, ie ansible.operatorframework.io. This would allow the adapter operators to version their own docs independently, and would mean that docs and code changes could happen in the same PR.

     * Could this be an intern project?

>    * How do versions work? netlify handles it

Reviewer 2: IMO the repos will be GO libs. IN this way, we should only have the Go docs which are gen automatically based in the comments made in the code. Also, we should ensure it by a lint check in the CI. IMO it would be done as it is for controller-runtime. E.g Lint : https://github.com/kubernetes-sigs/controller-runtime/pull/885 and docs https://godoc.org/sigs.k8s.io/controller-runtime

See that any change in the LIB which is, for example, a breaking change will force us to do changes in the SDK CLI tool which is using it and then, will force us to update the SDK website/guide docs. So, I do not think that it would have any downside over keep the website docs updated accordingly. 

WDYT @joelanford @asmacdo @jmrodri ?

Author: @camilamacedo86 yeah that makes sense regarding the go libs. They won't need the docs. Makes sense.

authors:
  - "@jesusr"
reviewers:
  - "@estroz"
  - "@hasbro17"
  - "@joelanford"
  - "@fabianvf"
  - "@shawn-hurley"
approvers:
  - "@estroz"
  - "@hasbro17"
  - "@joelanford"
  - "@fabianvf"
  - "@shawn-hurley"
creation-date: 2020-05-18
last-updated: 2020-05-27
status: implementable
see-also:
replaces:
superseded-by:
---

# Split SDK into multiple projects

## Release Signoff Checklist

- [X] Enhancement is `implementable`
- [X] Design details are appropriately documented from clear requirements
- [X] Test plan is defined
- [X] Graduation criteria for dev preview, tech preview, GA

## Open Questions [optional]

1. What do we do with other shared tooling, CI, test helpers, etc. (e.g.
   changelog generator)? Do we need a separate repo for that?
   * Answer: We should visit these on a case by case basis as we see what is needed
     during the splitting.

## Summary

This is the overarching enhancement that will plan how to split functionality
out of the current operator-sdk project into separate projects. The initial
thought on component breakdown mirrors the kubebuilder and controller-runtime
projects:

1. Common libraries and operator building blocks

> 1. Project scaffolding and the main SDK CLI

Reviewer 3: ```suggestion
1. Go scaffolding plugin and the main SDK CLI
```

better testing and easier to contribute to the smaller projects.

While there are a few complexities with having separate repos, like PR
coordination between repos, or just the working complexity of having to point
at different repos to pick up your dep changes, these are obstacles we can
overcome. They also aren’t that much different from other dependencies we
work with.

### Goals

Split the operator-sdk repo into 4 separate repos:

1. CLI and Go Operator
1. SDK library
1. Helm operator
1. Ansible operator

This SDK library split would be a requirement for a 1.0 release. The Ansible
and Helm operators could go afterwards if need be.

### Non-Goals

* We will not solve the downstream SDK in this enhancement.
* We will not solve all of the items required for a 1.0 release.
* Defining when version bumps occur in each repo.

## Proposal

Split the operator-sdk repo into 4 separate repos as outlined in the Goals
section above.

1. CLI and Go Operator
   * contains the Go operator code
   * contains the CLI code
   * related scaffolding
   * lives in operator-framework/operator-sdk
   * will have a deprecation policy

>      * we can deprecate in minor versions; removal becomes a major version bump

Reviewer 3: Is this referring to plugin version?

Author: The raw notes from the document were 
> do we need a deprecation policy, i.e. subcommand renamed; how many releases 1? 2?
we can deprecate in minor versions; removal becomes major versions
prior to 1.0 release, review entire CLI; axe all legacy support for existing project layout; 

I'll remove it from this because it is really a larger discussion with supporting the SDK and associated projects in the long term.

>    * this split is dependent on Kubebuilder Phase 2.

Reviewer 3: I don't think this is true, since Helm/Ansible plugins are phase 1 plugins.

Author: I made it phase 2 because I didn't want the moving of repos to affect phase 2 work since that might touch a couple repos vs one right now.

Author: I updated the comment to be a bit more clear:

> the split of the CLI/Go Operator repo is dependent on Kubebuilder Phase 2
to avoid creating a moving target.

And I added this to the helm and ansible splits respectively:
> the Helm plugin is dependent on Kubebuilder Phase 1.
> the Ansible plugin is dependent on Kubebuilder Phase 1.

1. SDK library
   * contains utility code like package status, status condition, annotated
     watcher, etc.
   * this is something that operator projects (including Helm and Ansible
     operators) would import
   * depends on controller-runtime

>    * lives in operator-framework/sdk-lib (or some other name TBD)

Reviewer 3: `operator-lib`? Thoughts?

Author: I like that better, makes it more operator centric than sdk centric.

   * this library should probably *not* be 1.0; it's most likely to change often
   * use Go apidiff tool to determine major version bumps
1. Helm operator
   * contains Helm operator code

>    * related scaffolding

Reviewer 3: ```suggestion
   * related plugin and scaffolding
```

   * lives in operator-framework/helm-operator
1. Ansible operator
   * contains Ansible operator code

>    * related scaffolding

Reviewer 3: ```suggestion
   * related plugin and scaffolding
```

   * lives in operator-framework/ansible-operator

![Image of project dependencies](split-sdk-into-multiple-projects-deps.png)

### Tasks

1. Triage pkg directory to determine what goes to which repo
1. Use git commands to extract directories or files with history into these
   new repos. Exact commands will be TBD.
1. Evaluate if making a separate repo for the hack directory is necessary or
   if we just copy it to each of the 4 repos.
1. CLI and Go Operator
   * keep operator-framework/operator-sdk
   * cmd, internal directories remain here
   * prior to 1.0 release, review entire CLI; axe all legacy support for
     existing project layout
   * consider making `generate csv` separate like `controller-gen` to support
     OLM's plan for CSVless bundles (could be done later)
1. SDK library
   * create new repo operator-framework/sdk-lib
   * export the directories / files based on task 1 triage.
1. Helm operator and plugin
   * create new repo operator-framework/helm-operator
   * export the pkg/helm tree to the new repo at the top level maintaining
     history

>      * Deprecate pkg/helm and replace with [Joe's helm repo](https://github.com/joelanford/helm-operator)

Reviewer 3: Can you elaborate on why/how we're replacing this package with @joelanford's repo?

Reviewer 4: Great question. Something that probably needs more discussion.

The intent of the new repo is to:
- Be fully backwards compatible with the existing helm operator implementation
- Re-organize the underlying libraries to actually make them expose a sane public API and make unit testing easier.
- Make it extensible so that others can add their own Go code customizations for hybrid use cases.
- Cleanup a bunch of issues that make the current implementation difficult to add new features and fix bugs while ensuring that things don't break.

In addition to the feature set of the current implementation, we also need to make sure we have all the necessary tooling to continue publishing the helm operator base image upstream and downstream for all of the required architectures.

Author: I've updated the text that this will be covered in a separate proposal.  I also updated the JIRA story for the helm-operator enhancement to contain this conversation.
